feat: Frontend for Bedrock Knowledge Base

coyanagiaws opened the pull request:
*Issue #, if available:*
https://github.com/aws-samples/bedrock-claude-chat/issues/440 (Base issue: https://github.com/aws-samples/bedrock-claude-chat/issues/139)

*Description of changes:*
In the cdk.json settings, the useBedrockKnowledgeBaseForRag: bool option will be used to allow the use of either Knowledge Base (KB) or pgvector, but not both.
By default, useBedrockKnowledgeBaseForRag is set to false.

The difference between the previous Bot Editor (with pgvector support) and the current one is:

- You can now add a S3 data source by inputting an S3 URI into the Knowledge section

   <img width="600" src="https://github.com/user-attachments/assets/be59bab6-a1fa-4981-8477-a877780ef4f0">

- In the Knowledge Detail Settings section, you can choose the following:

   - Embeddings model type (Titan Embeddings G1 - Text v1.2 / Embed Multilingual v3)
   - Chunking strategy (Default / Fixed-size / No chinking)
   - Document analyzer (Kuromoji / ICU)

   <img width="600" src="https://github.com/user-attachments/assets/89dfd051-a844-4646-b2d6-333edfa7d1a6">

- In the Search Settings, you can select the search type (Hybrid / Semantic) in addition to the number of max results.

   <img width="600" src="https://github.com/user-attachments/assets/794ca08f-6901-4329-9dbb-f6aebbfa3c9e">


#### Limitations of Bots created with the previous Bot Editor (with pgvector support):

The previously created bots become unsupported and are subject to the following limitations:

- Previously created Bots will be displayed with the prefix of [Unsupported, Read-only] in the Bot Console screen.

  <img width="600" src="https://github.com/user-attachments/assets/d1867431-c280-45a3-b81e-b0fdad60b18e">

- They cannot be edited.
- They cannot be shared, and they cannot be reverted to non-shared status.
- You cannot publish the API, create an API Key, or unpublish the API.

If you want to stop sharing or stop the published API, you will need to delete the Bot.

coyanagiaws committed:
implement BotEditPageV2 with KB supported

coyanagiaws committed:
addressed review comments

coyanagiaws committed:
revert start and copy link actions to be enabled

coyanagiaws committed:
fix eslint warning

coyanagiaws committed:
fix build fail

statefb commented on frontend/src/components/Slider.tsx:
Selectのdisableと同様、disable時はonhover時にカーソルを変更しないようお願いします！

coyanagiaws replied:
ありがとうございます。スタイルを cursor-default に変更するよう対応しました。[commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/631e8778869a5d7e35e9a7ab1500713102503620)

statefb commented on frontend/src/features/knowledgeBase/components/DescriptiveSelect.tsx:
こちらのコンポーネントは通常の汎用コンポーネントにしても良いかと思いました（feat/knowledgebaseではなく）。kb下に置いた特別な理由はありますか？

statefb replied:
お手間でなければstory化もお願いします🙇

coyanagiaws replied:
@statefb 汎用化していないのは、kb でしか現状利用しないものは feature 以下で閉じる、というポリシーに沿ったためですね。あと汎用化する場合、Select.tsx を拡張するか、単体で DescriptiveSelect.tsx とするかが決めきれなかったというのもあります。

statefb replied:
なるほどです。@wadabeeさんの意見も聞いてみたいです

coyanagiaws replied:
@wadabee さんと会話し、Select.tsx を拡張することになりました。こちらの commit にて対応しました。[commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/642445d4e20849cb8596e9d4172e39a399881e1f)

statefb commented on frontend/src/features/knowledgeBase/constants/index.ts:
```diff
@@ -0,0 +1,70 @@
+  } as OpenSearchParams,
+} as const;
+
+export const DEFAULT_OPENSEARCH_ANALYZER: {
```
以前お伝えしたところからブレてしまい申し訳ないのですが、Analyzerには何もセットしないオプションがあった方が良いかと思いました！

- デフォルト：None
- アジア圏 (中韓など) -> icu
- 日本 -> kuromoji

こちらでお願いします！

coyanagiaws replied:
@statefb None の場合は BE に渡すパラメーターは bedrockKnowledgeBase.openSearch.analyzer = null でよいでしょうか？

statefb replied:
はい、nullで大丈夫です！

coyanagiaws replied:
@statefb none のオプションを選択した場合は null を送信するように変更しました。（option変更が効かないバグがあったのでついでに修正済み）[commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/631e8778869a5d7e35e9a7ab1500713102503620) 

none の場合のオプション名と説明が適切か、念の為ご確認お願いします。

<img src="https://github.com/user-attachments/assets/c914328e-0ab5-4b84-8739-bf141018249c" width="600">

statefb replied:
ありがとうございます！OSの中で内部的にデフォルトのものがあるわけではなく、アナライザーを利用しない挙動となるので、下記のような文言のほうが良いかと思いましたが、いかがでしょう？（もっと良い表現があればお願いしますmm）
- title: アナライザーを利用しない
- description: アナライザーを利用しません

coyanagiaws replied:
OpenSearch の default analyzer を使うという意味のテキストに変更します

coyanagiaws replied:
@statefb "システム (OpenSearch) で定義されているデフォルトアナライザーを利用します。" に変更しました [commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/a8a21eb1c574768753ac29887203f887947d9355)

statefb committed:
add kb id

statefb commented on frontend/src/features/knowledgeBase/types/index.d.ts:
```diff
@@ -0,0 +1,42 @@
+export type BedrockKnowledgeBase = {
+  knowledgeBaseId?: string; // only present after bot is ready
```
すみません、Create / Modify時にknowlegeBaseIdも渡すようにしていただけないでしょうか？ボットを修正すると、この値がnullで上書きされてしまい動作しない不具合を見つけたためです（BEは[こちらのcommit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/599a01b3884fd39a51a7f8be1305327f87a4f283)で対応しています）

coyanagiaws replied:
@statefb create 時は ulid を送信し、modify では渡ってきた値をそのまま送信するよう修正しました [commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/631e8778869a5d7e35e9a7ab1500713102503620)

statefb replied:
細かくて恐縮ですが、kbIdはstep functionsの中でbedrockのサービス側で採番して振られるので、作成時はnullでお願いします！
```
const [knowledgeBaseId, setKnowledgeBaseId] = useState<string>(ulid());
```

coyanagiaws replied:
@statefb create 時は null を送信するよう修正しました [commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/a8a21eb1c574768753ac29887203f887947d9355)

statefb commented on frontend/src/features/knowledgeBase/pages/BotEditPageV2.tsx:
コンポーネント名のsuffixがV2ですとBrChat自体のバージョンとの齟齬など誤解を招きそうなので、たとえば`BotKbEditPage`、`BotKnowledgeBaseEditPage`など明示的にBedrock KBを利用するネーミングの方が良いかと思いましたが、いかがでしょうか？

statefb replied:
ボットはKnowledgeBaseだけでなくAgentも含むから、ということでしょうか。その意図ですとV2の方が良さそうですね（何か良い案思いついたらお伝えしますmm）

coyanagiaws replied:
@statefb この名前の意図としては、V2 で KB が optional から default に切り替わるという話だったので、エディターも V2 版と言って差し支えないのでは？と考えたため、ですね。とはいえ、V2 のタイミングで改めてコンポーネントのディレクトリ配置や名前を見直すとして、いったん現段階では feature で閉じた BotKbEditPage という名前が適切かもしれないですね。

statefb replied:
なるほどです。では`BotKbEditPage`でお願いします！

coyanagiaws replied:
@statefb 対応しました [commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/631e8778869a5d7e35e9a7ab1500713102503620)

statefb commented on frontend/src/features/knowledgeBase/pages/BotEditPageV2.tsx:
```diff
@@ -0,0 +1,1288 @@
+              ? ['']
+              : bot.knowledge.sourceUrls
+          );
+          setS3Urls(
```
簡単な正規表現チェックを入れていただけますか？(s3://で始まっている＆末尾が/で終わっている)。NGの場合はアラートを表示し、作成・更新ボタンをdisableする処理を入れてください！（オリジナルのEditPageと同等の仕様）

coyanagiaws replied:
@statefb ありがとうございます！こちらはすでに isValid 関数内で対応済みです。

statefb replied:
作成ボタンクリック時にエラーメッセージが出る仕様ですね。了解です！

statefb commented on frontend/src/features/knowledgeBase/pages/BotEditPageV2.tsx:
```diff
@@ -0,0 +1,1288 @@
+                          type="text"
+                          disabled={isLoading}
+                          value={s3Url}
+                          placeholder={'s3://example-bucket/'}
```
バケットだけでなくパスにも対応していることが明示できると親切かと思いました！

ex) s3://example-bucket/path/to/data-source/

coyanagiaws replied:
@statefb ありがとうございます、placeholder 修正しました [commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/631e8778869a5d7e35e9a7ab1500713102503620)

coyanagiaws committed:
addressed review comments
- file name change from 'BotEditPageV2' to 'BotKbEditPage'
- add none option for opensearch analyzer
- fix cursor style for Slider disabled state
- send knowledgeBaseId when creation and editing a bot
- change s3 url placeholder value

coyanagiaws committed:
commit forgotten changes

coyanagiaws committed:
enhance Select component to display description

statefb commented on frontend/src/@types/bot.d.ts:
```diff
@@ -45,6 +47,7 @@ export type BotSyncStatus = 'QUEUED' | 'RUNNING' | 'SUCCEEDED' | 'FAILED';
 
 export type BotListItem = BotMeta & {
   available: boolean;
+  hasBedrockKnowledgeBase?: boolean;
```
必ず存在する属性なので、optinoal除外ください！

coyanagiaws replied:
@statefb 対応しました！ [commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/a8a21eb1c574768753ac29887203f887947d9355)

fsatsuki commented on frontend/src/vite-env.d.ts:
```diff
@@ -13,6 +13,7 @@ interface ImportMetaEnv {
   readonly VITE_APP_CUSTOM_PROVIDER_ENABLED: string;
   readonly VITE_APP_CUSTOM_PROVIDER_NAME: string;
   readonly VITE_APP_ENABLE_MISTRAL: string;
+  readonly VITE_APP_ENABLE_KB: string;
```
.env.templateにも書いていただくとローカル開発がやりやすいのでお願いしたいです。

coyanagiaws replied:
@fsatsuki ！見落としていました、ありがとうございます。対応しました。[commit](https://github.com/aws-samples/bedrock-claude-chat/pull/509/commits/f4d7ee83e37173bd2d2895451ece6c01c7172f00)

statefb committed:
add: migration guide

wadabee commented on docs/migration/V1_TO_V2.md:
```diff
@@ -0,0 +1,25 @@
+
+## Overview
+
+The v2 update introduces a significant change by replacing pgvector on Aurora Serverless and ECS-based embedding with [Amazon Bedrock Knowledge Bases](https://docs.aws.amazon.com/bedrock/latest/userguide/knowledge-base.html). This change is not backward compatible. If you want to continue using pgvector, avoid upgrading to v2. If you plan to transition, start by updating to v1.3.0 and enabling th ...
```
v2にアップグレードするとAurora関連リソースが全て削除される旨も書いていたほうが良いかなと思いました。

statefb replied:
修正しました！

wadabee commented on docs/migration/V1_TO_V2.md:
```diff
@@ -0,0 +1,25 @@
+
+Replacement of pgvector and ECS Embedding: The previous system using pgvector on Aurora Serverless and ECS for embedding is replaced by Amazon Bedrock Knowledge Bases, which utilize an OpenSearch backend. This offers a fully managed solution, improving reliability and ease of maintenance.
+
+### Migration Process
+
+- Update to v1.3.0: Ensure you are using version 1.3.0 or later.
+- Enable Knowledge Bases: Set `useBedrockKnowledgeBasesForRag` to `true` in [cdk.json](../../cdk/cdk.json).
+- Read-Only Mode: Once enabled, your existing pgvector-based bots become read-only, preventing further edits but allowing continued use. Recommend to re-create another bot on Knowledge Bases.
```
1.3.0でKBを有効化すると、AuroraとKBの両方が課金される旨の記載があったほうが良いかなと思いました。

wadabee replied:
ボットの再作成手順もあると良いと思います。
「画面から新しいボット作成画面を開いて、pgvectorと同じ定義で再作成してね」というレベルで大丈夫と思います！

statefb replied:
修正しました！

coyanagiaws committed:
addressed review comments
- modify resource for opensearch
- send null knowlegeBaseId
- fix hasBedrockKnowledgeBase type

coyanagiaws committed:
addressed review comments
- add VITE_APP_ENABLE_KB in .env.template

statefb committed:
fix doc

coyanagiaws commented on docs/README_ja.md:
```diff
@@ -1,5 +1,8 @@
 # Bedrock Claude Chat
 
+> [!Warning]
+> 近々、v2 へのメジャーアップデートを予定しています。v1 とは後方互換性がなく、**既存の RAG ボットは使用できなくなる予定です。** 詳しくは[ガイド](./docs/migration/V1_TO_V2.md)をご参照ください。
```
リンク先が間違っているかも。正しくは `./migration/V1_TO_V2.md` ですかね。

statefb committed:
fix doc link

statefb commented:
LGTM!
